questionnaire_linking.py

```diff
@@ -0,0 +1,117 @@
+    global error_count
+    for line_number, questionnaire_link in enumerate(qid_link_file_reader, 1):
+        error_count = 0
+        for i in questionnaire_link:
+            questionnaire_link[i] = questionnaire_link[i].replace(" ", "")
+        validate_qid_len_and_type(line_number, questionnaire_link["qid"])
+        validate_qid_third_digit(line_number, questionnaire_link["qid"])
+        validate_check_digits(line_number, questionnaire_link["qid"])
+        case_id = validate_case_ref(line_number, questionnaire_link["case_ref"])
+
+        if error_count > 0:
+            print(f'Error: {questionnaire_link} on line {line_number} have FAILED')
+        else:
+            post_message_to_queue(case_id, questionnaire_link["qid"], line_number)
+        print("\n")
+
+
+def validate_case_ref(line_number, case_ref):
+    global error_count
+    if len(case_ref) != 8 or not case_ref.isdigit():
+        print(f'Error: Either wrong length or type located on Case Reference {case_ref}, line {line_number}')
+        error_count += 1
+    if error_count == 0:
+        response = requests.get(f'http://{Config.CASEAPI_HOST}:{Config.CASEAPI_PORT}/cases/ref/{case_ref}')
+        if response.status_code == 404:
+            print(f'Error: Case Reference {case_ref} on line {line_number} does not exist')
+            error_count += 1
+            return None
+        response.raise_for_status()
+        return response.json()["id"]
+
+
+def validate_qid_len_and_type(line_number, qid):
+    global error_count
+    if len(qid) != 16 or not qid.isdigit():
+        print(f'Error: Either wrong length or type located on QID {qid}, line {line_number}')
+        error_count += 1
+
+
+def validate_qid_third_digit(line_number, qid):
+    global error_count
+    if qid[2] != "2":
+        print(f'Error: QID {qid}, line {line_number} tranche identifier must be 2')
+        error_count += 1
+
+
+def validate_check_digits(line_number, qid):
+    global error_count
+    if not qid_checksum_validator.validate(qid, int(Config.QID_MODULUS), int(Config.QID_FACTOR))[0]:
+        print(f'Error: Check Digits incorrect for QID {qid}, line {line_number}')
+        error_count += 1
+
+
+def post_message_to_queue(case_id, qid, line_number):
+    message = {
+        "event": {
+            "type": "QUESTIONNAIRE_LINKED",
+            "source": "RM",
+            "channel": "RM",
+            "dateTime": datetime.utcnow().isoformat() + 'Z',
+            "transactionId": str(uuid.uuid4())
+        },
+        "payload": {
+            "uac": {
+                "questionnaireId": qid,
+                "caseId": case_id,
+            }
+        }
+    }
+    with RabbitContext() as rabbit:
+        rabbit.publish_message(json.dumps(message), "application/json", None, exchange='events',
+                               routing_key='event.questionnaire.update')
+        print(f'Case ID {case_id} and QID {qid} on line {line_number} have PASSED')
```

Comment:
it's more that they've been "linked" than "passed" - passed to me means  that validation was OK, but in actual fact we've issued a message here so we've done something more than just validation.

Reply:
Would this be better:

print(f'Line {line_number}: Case ID {case_id} and QID {qid} have PASSED and have been LINKED')